### PR TITLE
[object] Store package version in object

### PR DESCRIPTION
**Pull request by amnn**
Soon it will no longer be the case that a package's version is always `1`, because:

- Framework Upgrades will write new versions of the framework at increasing versions in the store.
- Package Upgrades will pun the package's version in storage with its version in the chain of upgrades (Even though subsequent versions will exist at distinct ObjectIDs).

This PR retires `PACKAGE_VERSION` and stores the package version in `MovePackage` instead (although initially, it will still always be 1).

This change also means that it is no longer possible to ascertain a package's version, when it is passed in as an input object.

## Test Plan

```
$ cargo nextest run
```

## Stack

- #7597

### CODE REVIEW

**Comment by amnn** [comment]
(Note, this is a stacked PR, so to review, just look at commits from `[object] Store package version in object` onwards, or wait to review until the bottom of the stack has landed.)

**Comment by amnn** [comment]
Test failure is probably legit, so let me move this back into draft mode.

**Comment by amnn** [comment]
I'm no longer sure that the issue with fullnode tests is due to this change.  @mystenmark, is this a known flaky test?  When run outside of `simtest`, I am seeing it sometimes succeeding, and sometimes not.

**Comment by amnn** [comment]
Thanks to @mystenmark I now know what's going on with the failing tests and they are legitimate after all.  Back to draft-mode to fix.

File: crates/sui-types/src/move_package.rs
Comment (tnowacki) [inline crates/sui-types/src/move_package.rs:48]: Just to double check from the summary of the PR, this is a pun for everything but the framework/stdlib correct? 
In other words, for non framework packages, we will never have two packages with the same id but different versions? 

Might be worth commenting here just since its different than how it works for MoveObjects

Comment (amnn) [reply crates/sui-types/src/move_package.rs:48]: Yep, that's correct.  Only the framework/stdlib can exist at multiple versions for the same ObjectID but all packages will store their package version (rather than the Lamport version of their parent transaction) in this field.

**Review by mystenmark — COMMENTED** [review]
This looks great to me - but I would like to have @mwtian review the transaction manager changes.

**Comment by mwtian** [comment]
Just curious about the background, is there a doc discussing what guarantees that a transaction will not be executed with different input package versions, and why it is necessary to support creating a package at an initial version other than 1?

File: crates/sui-core/src/authority/authority_store.rs
Comment (mwtian) [inline crates/sui-core/src/authority/authority_store.rs:339]: How many versions can there be on the same object ID, for MovePackage?

**Comment by amnn** [comment]
@mwtian, 

> Just curious about the background, is there a doc discussing what guarantees that a transaction will not be executed with different input package versions, and why it is necessary to support creating a package at an initial version other than 1?

This change is to support two kinds of upgrades:

- Framework Upgrades, where there might be multiple versions under the same Object ID (`0x1` or `0x2`), but new versions are only added on epoch boundaries.
- Third party package upgrades, where each new version of the package will exist at a fresh Object ID, but its version will be one greater than the previous version of the package, so there is only one version per object ID, but not necessarily version 1.

File: crates/sui-core/src/transaction_manager.rs
Comment (mwtian) [inline crates/sui-core/src/transaction_manager.rs:247]: Instead of probing missing_inputs table to see if the ObjectID is for package or object, I would prefer to be more exact here to help reasoning and debugging. How about:
1. Define an `InputKey` in authority_store that is basically `(pub ObjectID, pub Option<VersionNumber>)` or like `MissingInput` if you prefer.
2. Convert [1] in `authority.rs` to return `InputKey` instead of `ObjectKey`.
3. Send the `InputKey` to objects_committed()

[1]
```
let output_keys: Vec<_> = inner_temporary_store
            .written
            .iter()
            .map(|(_, ((id, seq, _), _, _))| ObjectKey(*id, *seq))
            .collect();
```

File: crates/sui-core/src/authority/authority_store.rs
Comment (mwtian) [inline crates/sui-core/src/authority/authority_store.rs:1220]: My preference would be to use `Option<VersionNumber>` for version which would be `None` for move packages. It may be easier to deal with but up to you.

File: crates/sui-core/src/authority/authority_store.rs
Comment (amnn) [reply crates/sui-core/src/authority/authority_store.rs:376]: (Resolving as covered in the main comment thread).

**Review by mwtian — APPROVED** [review]
LGTM for sui-core changes.

File: crates/sui-types/src/object.rs
Comment (lxfind) [inline crates/sui-types/src/object.rs:457]: Now that package version is mutable, should it still start from OBJECT_START_VERSION, or should it be the same as everything else based on lamport version?
Related, I wonder if OBJECT_START_VERSION is still used anywhere else? It not, at least we rename it to PACKAGE_START_VERSION if we don't want to apply lamport version to it yet

**Review by lxfind — COMMENTED** [review]
Do we need to include package object ref in the effects, now that we cannot tell its version based on the transaction data?

**Comment by mystenmark** [comment]
> Do we need to include package object ref in the effects, now that we cannot tell its version based on the transaction data?

Interesting question. This doesn't seem entirely necessary as a state commitment, because there can only exist one version of each unique package id (except for frameworks).

**Comment by lxfind** [comment]
>Interesting question. This doesn't seem entirely necessary as a state commitment, because there can only exist one version of each unique package id (except for frameworks).

Same argument applies to the shared_object_ids in the effects. So the question is whether we need it for other things, such as pruner, snapshot hash and etc..

File: crates/sui-types/src/object.rs
Comment (amnn) [reply crates/sui-types/src/object.rs:457]: I decided to keep `OBJECT_START_VERSION` because it is still used in a couple of places, for genesis and test objects.

I don't think packages should use lamport versions.  The reason to use lamport versions is related to interactions between child objects and wrapping/unwrapping, which packages don't participate in, so it's more valuable to retain a human comprehensible version for packages (i.e. if you see that the package is at version X, you know it has X antecedent versions).

To support that, I'll need to parametrise this field, so it isn't always `OBJECT_START_VERSION`, but I'll do that in a follow-up PR.  It's not clear what the best way to do that is without seeing how it will be used.

**Comment by amnn** [comment]
Let me kick off a thread on the question of package refs in effects -- I also think it probably won't be necessary, but I'm missing context on ancillary uses for effects.  I don't think that will affect the substance of this PR though, so let me land this to unblock the protocol upgrade work building on top of it.

**Pull request merged**
